File: src/test/java/cube8540/oauth/authentication/users/application/DefaultUserApprovalAuthorityServiceTest.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.users.domain.ApprovalAuthority;
import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.UserNotFoundException;
import cube8540.oauth.authentication.users.domain.UserRepository;
import cube8540.oauth.authentication.users.domain.UserValidatorFactory;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InOrder;

import java.util.Collection;
import java.util.Set;

import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.AUTHORITIES_A;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeComposeApprovalAuthorities;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeDefaultUser;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeEmptyUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeValidatorFactory;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.when;

@DisplayName("유저 기본 승인 권한 서비스")
public class DefaultUserApprovalAuthorityServiceTest {

    @Test
    @DisplayName("유저를 찾을 수 없을때 유저 승인 권한 검색")
    void getApprovalAuthoritiesWhenUserNotFound() {
        UserRepository repository = makeEmptyUserRepository();

        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        assertThrows(UserNotFoundException.class, () -> service.getApprovalAuthorities(RAW_USERNAME));
    }

    @Test
    @DisplayName("유저를 찾으면 그 유저의 승인 권한을 반환 해야 한다.")
    void ifUserFoundReturnsFoundedUsersApprovalAuthorities() {
        User user = makeDefaultUser();
        UserRepository repository = makeUserRepository(USERNAME, user);
        Set<ApprovalAuthority> approvalAuthorities = makeComposeApprovalAuthorities();
        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        when(user.getApprovalAuthorities()).thenReturn(approvalAuthorities);

        Collection<ApprovalAuthority> results = service.getApprovalAuthorities(RAW_USERNAME);
        assertEquals(approvalAuthorities, results);
    }

    @Test
    @DisplayName("찾을 수 없는 유저에게 승인 권한을 추가 했을시")
    void grantApprovalAuthoritiesToNotFoundUser() {
        UserRepository repository = makeEmptyUserRepository();
        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        assertThrows(UserNotFoundException.class, () -> service.grantApprovalAuthorities(RAW_USERNAME, AUTHORITIES_A));
    }

    @Test
    @DisplayName("승인 권한을 추가")
    void grantNotAllowApprovalAuthorities() {
        User user = makeDefaultUser();
        UserValidatorFactory factory = makeValidatorFactory();
        UserRepository repository = makeUserRepository(USERNAME, user);
        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        service.setValidatorFactory(factory);

        service.grantApprovalAuthorities(RAW_USERNAME, AUTHORITIES_A);
        InOrder inOrder = inOrder(user, repository);
        AUTHORITIES_A.forEach(authority -> inOrder.verify(user)
                .addApprovalAuthority(authority.getClientId(), authority.getScopeId()));
        inOrder.verify(user).validation(factory);
        inOrder.verify(repository).save(user);
    }

    @Test
    @DisplayName("찾을 수 없는 유저의 승인 권한 삭제")
    void revokeApprovalAuthoritiesToNotFoundUser() {
        UserRepository repository = makeEmptyUserRepository();
        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        assertThrows(UserNotFoundException.class, () -> service.revokeApprovalAuthorities(RAW_USERNAME, AUTHORITIES_A));
    }

    @Test
    @DisplayName("승인 권한 삭제")
    void revokeApprovalAuthorities() {
        User user = makeDefaultUser();
        UserRepository repository = makeUserRepository(USERNAME, user);
        DefaultUserApprovalAuthorityService service = new DefaultUserApprovalAuthorityService(repository);

        service.revokeApprovalAuthorities(RAW_USERNAME, AUTHORITIES_A);
        InOrder inOrder = inOrder(user, repository);
        AUTHORITIES_A.forEach(authority -> inOrder.verify(user)
                .revokeApprovalAuthority(authority.getClientId(), authority.getScopeId()));
        inOrder.verify(repository).save(user);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/user/application/DefaultUserApprovalAuthorityServiceTest.kt
package cube8540.oauth.authentication.user.application

import cube8540.oauth.authentication.users.application.DefaultUserApprovalAuthorityService
import cube8540.oauth.authentication.users.domain.*
import io.mockk.every
import io.mockk.mockk
import io.mockk.verifyOrder
import java.util.Optional
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class DefaultUserApprovalAuthorityServiceTest {
    private val repository: UserRepository = mockk()
    private val validatorFactory: UserValidatorFactory = mockk()
    private val service = DefaultUserApprovalAuthorityService(repository)

    init {
        every { repository.save(any()) } returnsArgument 0
        service.validatorFactory = validatorFactory
    }

    @Nested
    inner class GetApprovalAuthoritiesTest {

        @Test
        fun `request user is not registered in repository`() {
            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.getApprovalAuthorities("username") }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val userApprovalAuthorities = mutableSetOf(
                ApprovalAuthority("clientA", "SCOPE-1"),
                ApprovalAuthority("clientB", "SCOPE-2"),
                ApprovalAuthority("clientC", "SCOPE-3")
            )
            val user: User = mockk {
                every { approvalAuthorities } returns userApprovalAuthorities
            }

            every { repository.findById(Username("username")) } returns Optional.of(user)

            val result = service.getApprovalAuthorities("username")
            assertThat(result).isEqualTo(userApprovalAuthorities)
        }
    }

    @Nested
    inner class GrantApprovalAuthorityTest {

        @Test
        fun `request user is not registered in repository`() {
            val requestApprovalAuthorities: Collection<ApprovalAuthority> = listOf(
                ApprovalAuthority("clientA", "SCOPE-1"),
                ApprovalAuthority("clientB", "SCOPE-2"),
                ApprovalAuthority("clientC", "SCOPE-3")
            )

            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.grantApprovalAuthorities("username", requestApprovalAuthorities) }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val requestApprovalAuthorities: Collection<ApprovalAuthority> = listOf(
                ApprovalAuthority("clientA", "SCOPE-1"),
                ApprovalAuthority("clientB", "SCOPE-2"),
                ApprovalAuthority("clientC", "SCOPE-3")
            )
            val user: User = mockk(relaxed = true)

            every { repository.findById(Username("username")) } returns Optional.of(user)

            service.grantApprovalAuthorities("username", requestApprovalAuthorities)
            verifyOrder {
                user.addApprovalAuthority("clientA", "SCOPE-1")
                user.addApprovalAuthority("clientB", "SCOPE-2")
                user.addApprovalAuthority("clientC", "SCOPE-3")
                user.validation(validatorFactory)
                repository.save(user)
            }
        }
    }

    @Nested
    inner class RevokeApprovalAuthorityTest {

        @Test
        fun `request user is not registered in repository`() {
            val requestApprovalAuthorities: Collection<ApprovalAuthority> = listOf(
                ApprovalAuthority("clientA", "SCOPE-1"),
                ApprovalAuthority("clientB", "SCOPE-2"),
                ApprovalAuthority("clientC", "SCOPE-3")
            )

            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.revokeApprovalAuthorities("username", requestApprovalAuthorities) }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val requestApprovalAuthorities: Collection<ApprovalAuthority> = listOf(
                ApprovalAuthority("clientA", "SCOPE-1"),
                ApprovalAuthority("clientB", "SCOPE-2"),
                ApprovalAuthority("clientC", "SCOPE-3")
            )
            val user: User = mockk(relaxed = true)

            every { repository.findById(Username("username")) } returns Optional.of(user)

            service.revokeApprovalAuthorities("username", requestApprovalAuthorities)
            verifyOrder {
                user.revokeApprovalAuthority("clientA", "SCOPE-1")
                user.revokeApprovalAuthority("clientB", "SCOPE-2")
                user.revokeApprovalAuthority("clientC", "SCOPE-3")
                repository.save(user)
            }
        }
    }
}